app/src/main/java/com/globant/onboardingcalculatorwithkotlin/mvp/presenter/CalculatorPresenter.kt

```diff
@@ -62,17 +73,63 @@ class CalculatorPresenter(
             } else if (!model.first_operand.contains(DECIMAL_POINT)) {
                 model.first_operand = "${model.first_operand}$DECIMAL_POINT"
             } else {
-                view.showDecimalError()
+                view.showMessage(CalculatorError.DECIMAL_ERROR)
             }
         } else if (model.operator != EMPTY_CHAR) {
             if (model.second_operand.isEmpty()) {
                 model.second_operand = "$NUMBER_ZERO$DECIMAL_POINT"
             } else if (!model.second_operand.contains(DECIMAL_POINT)) {
                 model.second_operand = "${model.second_operand}$DECIMAL_POINT"
             } else {
-                view.showDecimalError()
+                view.showMessage(CalculatorError.DECIMAL_ERROR)
             }
         }
         updateVisor()
     }
+
+    private fun validOperation(): Boolean {
+        if (!model.second_operand.equals(NUMBER_ZERO))
+            return true
+        return false
+    }
+
+    private fun calculate(): String = when {
+        model.operator === OPERATOR_PLUS -> model.first_operand.toDouble() + model.second_operand.toDouble()
+        model.operator === OPERATOR_SUBSTRACTION -> model.first_operand.toDouble() - model.second_operand.toDouble()
+        model.operator === OPERATOR_MULTIPLY -> model.first_operand.toDouble() * model.second_operand.toDouble()
+        model.operator === OPERATOR_DIVIDE -> {
+            if (validOperation()) {
+                model.first_operand.toDouble() / model.second_operand.toDouble()
+            } else {
+                view.showMessage(CalculatorError.MATH_ERROR)
+                onClearButtonPressed()
+                EMPTY_STRING
+            }
+        }
+        else -> EMPTY_STRING
+    }.toString()
+
+    override fun onEqualPressed() {
+        if ((model.first_operand.isNotEmpty()) && (model.second_operand.isNotEmpty()) && (model.operator != EMPTY_CHAR)) {
+            model.first_operand = calculate()
+            model.result = model.first_operand
+            model.second_operand = EMPTY_STRING
+            model.operator = EMPTY_CHAR
+            if(model.first_operand.isNotEmpty()){
+                view.refreshVisor(model.result)
+            }else{
+                view.clearVisor()
+            }
+        } else {
+            view.showMessage(CalculatorError.INCOMPLETE_OPERATION)
+        }
+    }
+}
+
+enum class CalculatorError(val message: Int) {
```

Comment:
this value (val message: Int) its neccesary?

Reply:
Refactored!